PR: https://github.com/odoo/odoo/pull/96134

From: d8d47f9ff8554f4b39487fd2f13c153c7d6f958d
From: william-andre
Date: 2022-08-03 11:44:49

Breaking data model changes score: 1, change matches:
-    is_anglo_saxon_line = fields.Boolean()

Total Changes: 66

[REF] accounting v16. Yeeeeaah

TLDR:
* invoices are implemented using computed methods instead of onchange
* the synchronization only happens when switching tabs in the Form view
  to improve perfs.

_______________________________________________________________________

The whole engine of the synchronization of Invoices to the Journal
Entries has been refactored
* by using computed fields instead of onchange functions
* by synchronizing only from invoice to journal entry in `create` and
  `write`
* by saving when switching tabs on the Invoice form, to synchronize
  before showing the values

This comes with numerous advantages:
* no need to call the onchange methods manually
* no need to use the Form emulator to build invoices (i.e. EDI, OCR,
  intercompany, ...)
* the performance for invoices with many lines improves drastically, going
  from 2 minutes to 4 seconds to create an invoice with 500 lines
* the model is more declarative, we can now see how the values are computed
  instead of having the values being copied from various places.
* remove the hack in `onchange` that disabled the recursivity of it,
  which was unexpected and needed to be managed manually in all the
  onchange methods

This means that:
* Some fields need to be exclusively computed on journal entries values
  or invoice values, more specifically the Tax Summary widget.
  It is now
    - computed from entry lines, when opening the view
    - computed from invoice lines when changing those, because the tax lines
      will need to be recomputed anyways, erasing previously set values
    - set with an inverse function when saving; after the sync has been done
* Some possible operations previously possible have been dropped.
  (i.e. look at the removed test test_in_invoice_line_onchange_accounting_fields_1)
  This is because such a behavior was undefined (how is changing the balance going
  to affect the unit price? How is the amount currency going to affect it?)

_______________________________________________________________________

Implementation Details
----------------------

The "dynamic lines", meaning the payment terms and the tax lines are now
only created in the `create` and `write` functions.
In order to reduce code duplication, it has been implemented using
context managers used in both `account.move` and `account.move.line`
These context managers help comparing the values before/after, acting
like a local `onchange`, but getting benefit from the dirty flags from
the `compute` dependences.
This is relying on computed fields on the move (`needed_terms`) and on
the lines (`compute_all_tax`) which contain the values needed for the
related move.
Depending on the needed values and the existing values (`term_key` and
`tax_key`, respectively) the context manager will determine what needs
to be created/updated/deleted.

Some related changes are to produce a `dict` instead of a `str` for the
`tax_totals` (previously `tax_totals_json`) fields, by simplicity to
reduce the complexity of IO, and simplicity of debugging, because the
logic of the field needed to change (cannot be computed at the same time
anymore since it needed the lines to be synced)

By simplicity, and also because it makes more sense, some boolean fields
have been merged into `display_type`:
* `is_rounding_line`
* `exclude_from_invoice_tab`
* `is_anglo_saxon_line`

The `price_unit`, `quantity` and other "invoice fields" are now not set
anymore on lines that are not product lines since it didn't make any
sense to have it.

Performances
------------

You have to keep in mind that a simple `create` didn't compute a lot of
fields, for instance not taxes were set, no payment terms,...
Now it does.

```python
import random
from timeit import timeit
from odoo import Command
domain = [('company_id', 'in', (False, self.env.company.id))]
products = self.env['product.product'].search(domain).ids
partners = self.env['res.partner'].search(domain).ids
taxes = self.env['account.tax'].search(domain).ids
def create(nmove, nline):
    self.env['account.move'].create([
        {
            'move_type': 'out_invoice',
            'partner_id': random.choice(partners),
            'invoice_line_ids': [
                Command.create({
                    'name': f'line{i}',
                    'product_id': random.choice(products),
                    'tax_ids': [Command.set([random.choice(taxes)])],
                })
                for i in range(nline)
            ]
        }
        for j in range(nmove)
    ])
                                                             # After  | Before
print(timeit("create(1, 1)", globals=globals(), number=1))   # 0.11   | 0.09
print(timeit("create(100, 1)", globals=globals(), number=1)) # 2.76   | 2.50
print(timeit("create(500, 1)", globals=globals(), number=1)) # 14.56  | 12.34
print(timeit("create(1, 100)", globals=globals(), number=1)) # 1.03   | 5.52
print(timeit("create(1, 500)", globals=globals(), number=1)) # 3.99   | 125.02
print(timeit("create(50, 50)", globals=globals(), number=1)) # 19.44  | 79.55
```

Another metric that can be used is running the test suite with
`--test-tags=/account` (only `account` installed)
* before: 404s, 267127 queries (366 tests)
* after: 318s, 232125 queries (362 tests)

Why this commit title?
----------------------

Someone told me that this was the perfect way of naming your commits.
c04065abd8f62c9a211c8fa824f5eecf68e61b73

task-2711317

closes odoo/odoo#96134

Related: odoo/upgrade#3715
Related: odoo/enterprise#29758
Signed-off-by: Laurent Smet <las@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock_account/models/account_move.py
+++ b/addons/stock_account/models/account_move.py
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 
-from odoo import fields, models
+from odoo import fields, models, api
 
 
 class AccountMove(models.Model):
@@ -15,15 +15,7 @@ class AccountMove(models.Model):
 
     def _get_lines_onchange_currency(self):
         # OVERRIDE
-        return self.line_ids.filtered(lambda l: not l.is_anglo_saxon_line)
-
-    def _reverse_move_vals(self, default_values, cancel=True):
-        # OVERRIDE
-        # Don't keep anglo-saxon lines if not cancelling an existing invoice.
-        move_vals = super(AccountMove, self)._reverse_move_vals(default_values, cancel=cancel)
-        if not cancel:
-            move_vals['line_ids'] = [vals for vals in move_vals['line_ids'] if not vals[2]['is_anglo_saxon_line']]
-        return move_vals
+        return self.line_ids.filtered(lambda l: l.display_type != 'cogs')
 
     def copy_data(self, default=None):
         # OVERRIDE
@@ -34,7 +26,7 @@ class AccountMove(models.Model):
             for copy_vals in res:
                 if 'line_ids' in copy_vals:
                     copy_vals['line_ids'] = [line_vals for line_vals in copy_vals['line_ids']
-                                             if line_vals[0] != 0 or not line_vals[2].get('is_anglo_saxon_line')]
+                                             if line_vals[0] != 0 or line_vals[2].get('display_type') != 'cogs']
 
         return res
 
@@ -59,7 +51,7 @@ class AccountMove(models.Model):
         res = super(AccountMove, self).button_draft()
 
         # Unlink the COGS lines generated during the 'post' method.
-        self.mapped('line_ids').filtered(lambda line: line.is_anglo_saxon_line).unlink()
+        self.mapped('line_ids').filtered(lambda line: line.display_type == 'cogs').unlink()
         return res
 
     def button_cancel(self):
@@ -69,7 +61,7 @@ class AccountMove(models.Model):
         # Unlink the COGS lines generated during the 'post' method.
         # In most cases it shouldn't be necessary since they should be unlinked with 'button_draft'.
         # However, since it can be called in RPC, better be safe.
-        self.mapped('line_ids').filtered(lambda line: line.is_anglo_saxon_line).unlink()
+        self.mapped('line_ids').filtered(lambda line: line.display_type == 'cogs').unlink()
         return res
 
     # -------------------------------------------------------------------------
@@ -128,7 +120,7 @@ class AccountMove(models.Model):
                 # Compute accounting fields.
                 sign = -1 if move.move_type == 'out_refund' else 1
                 price_unit = line._stock_account_get_anglo_saxon_price_unit()
-                balance = sign * line.quantity * price_unit
+                amount_currency = sign * line.quantity * price_unit
 
                 # Add interim account line.
                 lines_vals_list.append({
@@ -139,11 +131,10 @@ class AccountMove(models.Model):
                     'product_uom_id': line.product_uom_id.id,
                     'quantity': line.quantity,
                     'price_unit': price_unit,
-                    'debit': balance < 0.0 and -balance or 0.0,
-                    'credit': balance > 0.0 and balance or 0.0,
+                    'amount_currency': -amount_currency,
                     'account_id': debit_interim_account.id,
-                    'exclude_from_invoice_tab': True,
-                    'is_anglo_saxon_line': True,
+                    'display_type': 'cogs',
+                    'tax_ids': [],
                 })
 
                 # Add expense account line.
@@ -155,13 +146,12 @@ class AccountMove(models.Model):
                     'product_uom_id': line.product_uom_id.id,
                     'quantity': line.quantity,
                     'price_unit': -price_unit,
-                    'debit': balance > 0.0 and balance or 0.0,
-                    'credit': balance < 0.0 and -balance or 0.0,
+                    'amount_currency': amount_currency,
                     'account_id': credit_expense_account.id,
                     'analytic_account_id': line.analytic_account_id.id,
                     'analytic_tag_ids': [(6, 0, line.analytic_tag_ids.ids)],
-                    'exclude_from_invoice_tab': True,
-                    'is_anglo_saxon_line': True,
+                    'display_type': 'cogs',
+                    'tax_ids': [],
                 })
         return lines_vals_list
 
@@ -218,21 +208,19 @@ class AccountMove(models.Model):
 class AccountMoveLine(models.Model):
     _inherit = 'account.move.line'
 
-    is_anglo_saxon_line = fields.Boolean()
-
-    def _get_computed_account(self):
-        # OVERRIDE to use the stock input account by default on vendor bills when dealing
-        # with anglo-saxon accounting.
-        self.ensure_one()
-        self = self.with_company(self.move_id.journal_id.company_id)
-        if self.product_id.type == 'product' \
-            and self.move_id.company_id.anglo_saxon_accounting \
-            and self.move_id.is_purchase_document():
-            fiscal_position = self.move_id.fiscal_position_id
-            accounts = self.product_id.product_tmpl_id.get_product_accounts(fiscal_pos=fiscal_position)
+    def _compute_account_id(self):
+        super()._compute_account_id()
+        input_lines = self.filtered(lambda line: (
+            line.product_id.type == 'product'
+            and line.move_id.company_id.anglo_saxon_accounting
+            and line.move_id.is_purchase_document()
+        ))
+        for line in input_lines:
+            line = line.with_company(line.move_id.journal_id.company_id)
+            fiscal_position = line.move_id.fiscal_position_id
+            accounts = line.product_id.product_tmpl_id.get_product_accounts(fiscal_pos=fiscal_position)
             if accounts['stock_input']:
-                return accounts['stock_input']
-        return super(AccountMoveLine, self)._get_computed_account()
+                line.account_id = accounts['stock_input']
 
     def _eligible_for_cogs(self):
         self.ensure_one()
@@ -242,8 +230,12 @@ class AccountMoveLine(models.Model):
         self.ensure_one()
         if not self.product_id:
             return self.price_unit
-        original_line = self.move_id.reversed_entry_id.line_ids.filtered(lambda l: l.is_anglo_saxon_line
+        original_line = self.move_id.reversed_entry_id.line_ids.filtered(lambda l: l.display_type == 'cogs'
             and l.product_id == self.product_id and l.product_uom_id == self.product_uom_id and l.price_unit >= 0)
         original_line = original_line and original_line[0]
         return original_line.price_unit if original_line \
             else self.product_id.with_company(self.company_id)._stock_account_get_anglo_saxon_price_unit(uom=self.product_uom_id)
+
+    @api.onchange('product_id')
+    def _inverse_product_id(self):
+        super(AccountMoveLine, self.filtered(lambda l: l.display_type != 'cogs'))._inverse_product_id()
